src/main/java/org/broadinstitute/hellbender/engine/MultiVariantDataSource.java

```diff
@@ -0,0 +1,408 @@
+    {
+        if (targetSequence.getSequenceLength() != sourceSequence.getSequenceLength()) {
+            final String msg = String.format("Incompatible sequence lengths found (%s: %d) and (%s: %d)",
+                    sourceSequence.getSequenceName(),
+                    sourceSequence.getSequenceLength(),
+                    targetSequence.getSequenceName(),
+                    targetSequence.getSequenceLength());
+            throw new UserException.IncompatibleSequenceDictionaries(
+                    msg,
+                    sourceDataSourceName,
+                    sourceDictionary,
+                    targetDataSourceName,
+                    targetDictionary
+            );
+        } else if (!Objects.equals(targetSequence.getMd5(), sourceSequence.getMd5())) {
+            final String msg = String.format("Incompatible sequence MD5 values found (%s: %s) and (%s: %s)",
+                    sourceSequence.getSequenceName(),
+                    sourceSequence.getMd5(),
+                    targetSequence.getSequenceName(),
+                    targetSequence.getMd5());
+            throw new UserException.IncompatibleSequenceDictionaries(
+                    msg,
+                    sourceDataSourceName,
+                    sourceDictionary,
+                    targetDataSourceName,
+                    targetDictionary
+            );
+        }
+    }
+
+    /**
+     * Close the iterator currently open over the data sources, if there is one.
+     */
+    private void closeOpenIterationIfNecessary() {
+        if (currentIterator != null) {
+            currentIterator.close();
+            currentIterator = null;
+        }
+    }
+
+    /**
+     * Wrap the sourceIterator in a CloseableIterator to make it usable as a MergingIterator source.
+     */
+    private CloseableIterator<VariantContext> getCloseableIteratorWrapper(final Iterator<VariantContext> sourceIterator) {
+        validateFeatureSources();
+        Utils.nonNull(sourceIterator);
+
+        return new CloseableIterator<VariantContext>() {
+            Iterator<VariantContext> delegateIterator = sourceIterator;
+            @Override
+            public void close() { delegateIterator = null; }
+
+            @Override
+            public boolean hasNext() {
+                return delegateIterator.hasNext();
+            }
+
+            @Override
+            public VariantContext next() {
+                return delegateIterator.next();
+            }
+        };
+    }
+
+    /**
+     * Invalidate the cached VCFHeader and SAMSequenceDictionary whenever the set of backing FeatureDataSources
+     * changed in order to force them to be recreated and cached on demand.
+     */
+    private void invalidateCachedHeaderAndDictionary() {
+        mergedHeader = null;
+        mergedDictionary = null;
+    }
+
+    /**
+     * Validate the requirement that one or more variant data sources have been added
+     */
+    private void validateFeatureSources() {
```

Comment:
This can probably also be deleted if adding data sources is done only at construction.